test/Dialect/SystemC/module-error.mlir

```diff
@@ -382,3 +382,99 @@ systemc.module @variableNameCollision () {
+
+// -----
+
+systemc.module @submodule () { }
+
+hw.module @verilatedCannotReferenceNonHWModule() {
+  // expected-error @+1 {{symbol reference 'submodule' isn't a module}}
+  systemc.interop.verilated "verilated" @submodule () -> ()
+}
+
+// -----
+
+hw.module @referenceNonExistentSymbol() {
+  // expected-error @+1 {{Cannot find module definition 'submodule'}}
+  systemc.interop.verilated "verilated" @submodule () -> ()
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule (%arg0: i32) -> ()
+
+hw.module @wrongNumberOfOperands() {
+  // expected-error @+1 {{op has a wrong number of operands; expected 1 but got 0}}
+  systemc.interop.verilated "verilated" @submodule () -> ()
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule (%arg0: i32) -> ()
+
+hw.module @wrongNumberOfInputNames(%arg0: i32) {
+  // expected-error @+1 {{has a wrong number of input port names; expected 1 but got 0}}
+  "systemc.interop.verilated"(%arg0) {instanceName="verilated", moduleName=@submodule, inputNames=[], resultNames=[]} : (i32) -> ()
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule (%arg0: i32) -> ()
+
+hw.module @wrongOperandType(%arg0: i8) {
+  // expected-error @+1 {{operand type #0 must be 'i32', but got 'i8'}}
+  systemc.interop.verilated "verilated" @submodule (arg0: %arg0: i8) -> ()
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule (%arg0: i32) -> ()
+
+hw.module @wrongOperandLabel(%arg0: i32) {
+  // expected-error @+1 {{input label #0 must be "arg0", but got "a"}}
+  systemc.interop.verilated "verilated" @submodule (a: %arg0: i32) -> ()
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule () -> (out0: i32)
+
+hw.module @wrongNumResults() {
+  // expected-error @+1 {{has a wrong number of results; expected 1 but got 0}}
+  systemc.interop.verilated "verilated" @submodule () -> ()
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule () -> (out0: i32)
+
+hw.module @wrongResultType() {
+  // expected-error @+1 {{result type #0 must be 'i32', but got 'i8'}}
+  %verilated.out0 = systemc.interop.verilated "verilated" @submodule () -> (out0: i8)
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule () -> (out0: i32)
+
+hw.module @wrongNumberOfResultNames() {
+  // expected-error @+1 {{has a wrong number of results port labels; expected 1 but got 0}}
+  "systemc.interop.verilated"() {instanceName="verilated", moduleName=@submodule, inputNames=[], resultNames=[]} : () -> i32
+}
+
+// -----
+
+// expected-note @+1 {{module declared here}}
+hw.module.extern @submodule () -> (out0: i32)
+
+hw.module @wrongResultLabel() {
+  // expected-error @+1 {{input label #0 must be "out0", but got "o"}}
+  %verilated.out0 = systemc.interop.verilated "verilated" @submodule () -> (o: i32)
+}
```

Comment:
It is nice that all this behavior just falls out of the instance op refactor. 👍 One question will be do we need to test this, given that the verification is relying on shared utilities.

Reply:
We don't need it for statement/branch coverage, but it might be good to have one test per library call to test that everything is called correctly?

Reply:
I moved three of the tests to the HW errors test file because it didn't test for these three error messages before (such that the instance impl library is tested in one central place). I left one test for `systemc.interop.verilated` to check that the `verifySymbolUses` function calls the library and removed the remaining tests to reduce maintenance burden.